Diff API in Change Tracking (#914)
cap-js/docs: java/cds-data.md

:::

In case one of the images is empty, the `CdsDiffProcessor` traverses through the existing state treating it as an addition or removal mirroring the logic accordingly.

Changes detected by `CdsDiffProcessor` are reported to one or more visitors implementing the interface `CdsDiffProcessor.DiffVisitor`.

The visitor is added to `CdsDiffProcessor` with the `add()` method before starting the processing.

```java
diff.add(new DiffVisitor() {
  @Override
  public void changed(Path newPath, Path oldPath, CdsElement element, Object newValue, Object oldValue) {
      // changes
  }

  @Override
  public void added(Path newPath, Path oldPath, CdsElement association, Map<String, Object> newValue) {
      // additions
  }

  @Override
  public void removed(Path newPath, Path oldPath, CdsElement association, Map<String, Object> oldValue) {
      // removals
  }
});
```

The visitor can be added together with the [element filter](/java/cds-data#element-filters) that limits the subset of changes reported to the visitor.

```java
diff.add(
  new Filter() {
    @Override
    public boolean test(Path path, CdsElement element, CdsType type) {

>         return true;

Author: Below I have one.

   (path, element, isNull) -> isNull ? null : randomUUID()); // generator
```

## Diff Processor

In business logic, there might be the need to compare two states of an entity e.g. before and after an operation. Usually, you might want to react on changed values, for instance, to track changes accordingly.
To do that you can use `CdsDiffProcessor`, similar to the [Data Processor](/java/cds-data#cds-data-processor).
It traverses through two states of the entity and reports differences between them along the way.

Create an instance of the `CdsDiffProcessor` using the `create()` method:

```java
CdsDiffProcessor diff = CdsDiffProcessor.create();
```

You can compare the data represented as [structured data](/java/cds-data#structured-data), results of the CQN statements or arguments of event handlers.

> To do a comparison, `CdsDiffProcessor` requires the following in the data maps to compare:

Reviewer: ```suggestion
To do a comparison with the `CdsDiffProcessor`, the data maps to be compared need to adhere to the following requirements:
```

> - entities must include full set of primary keys
> - names of the elements must match the elements of the entity type
> - associations must be represented as [nested structures and associations](/java/cds-data#nested-structures-and-associations) according to the associations` cardinalities.

Reviewer: ```suggestion
- the data map must include values for all key elements
- the names in the data map must match the elements of the entity
- associations must be represented as [nested structures and associations](/java/cds-data#nested-structures-and-associations) according to the associations` cardinalities
```

The [delta representation](/java/working-with-cql/query-api#deep-update-delta) of collections is also supported.
Results of the CQN statements fulfill these conditions if the type [that comes with the result](/java/working-with-cql/query-execution#introspecting-the-row-type) is used instead of the entity type.

> You run the comparison by calling the `process()` method and supplying new and old image of the data as a `Map` (or a collection of them) and the type of the compared entity:

Reviewer: ```suggestion
To run the comparison, call the `process()` method and provide new and old state of the data as a `Map` (or a collection of them) and the type of the compared entity:
```

```java

> List<Map<String, Object>> newImage;

Reviewer: Image -> State or use Image instead of stage in the decription

List<Map<String, Object>> oldImage;
CdsStructuredType type;

diff.process(newImage, oldImage, type);
```

```java
Result newImage = service.run(Select.from(...));
Result oldImage = service.run(Select.from(...));

diff.process(newImage, oldImage, newImage.rowType());
```

> :::tip Draft-enabled Entities

Reviewer: ```suggestion
:::tip Comparing Draft-enabled Entities
```

> For draft-enabled entities, you may omit value of `IsActiveEntity` in the images.
> If you compare active and inactive state of the same entity using them as an old and new image make sure that
> the values of `IsActiveEntity` is either absent or the same in both images.

Reviewer: ```suggestion
If you compare the active with the inactive state of a draft-enabled entity, make sure that the `IsActiveEntity` values are either absent or the same in both images.
```

  }

  @Override
  public void added(Path newPath, Path oldPath, CdsElement association, Map<String, Object> newValue) {
      // additions
  }

  @Override
  public void removed(Path newPath, Path oldPath, CdsElement association, Map<String, Object> oldValue) {
      // removals
  }
});
```

The visitor can be added together with the [element filter](/java/cds-data#element-filters) that limits the subset of changes reported to the visitor.

```java
diff.add(
  new Filter() {
    @Override
    public boolean test(Path path, CdsElement element, CdsType type) {
        return true;
    }
  },
  new DiffVisitor() {
    ...
  }
);
```

You may add as many visitors as you need by chaining the `add()` calls.
Each instance of the `CdsDiffProcessor` can have own set of visitors added to it.

If your visitors need to be stateful, prefer one-time disposable objects for them. `CdsDiffProcessor` does not manage the state of them.

All values are compared using the standard Java `equals()` method including the elements with structured and arrayed types.

### Implementing a DiffVisitor

Additions and removals in the entity state reported as calls to the methods `added()` or `removed()`.

> They always receive complete added or removed state for the entity or an association.
> It is never traversed element by element.

Reviewer: ```suggestion
The called methods always receive the complete added or removed state for the entity or an association.
```

The methods `added()` and `removed()` have the following arguments:

- `newPath` and the `oldPath` as an instances of [`Path`](https://www.javadoc.io/doc/com.sap.cds/cds4j-api/latest/com/sap/cds/ql/cqn/Path.html) reflecting the new and old state of the entity.
- `association` as an instance of [`CdsElement`](https://www.javadoc.io/doc/com.sap.cds/cds4j-api/latest/com/sap/cds/reflect/CdsElement.html) given that the association is present.
- state of the changed data as a `Map` as the `newValue` or `oldValue`.

> The instances of the `Path` represent the placement of the changed item within the whole entity as a prefix to the state that is either added or removed. While these paths are always have the same structure, `oldPath` or `newPath` respectively, could have empty values that represent the absence of the data.

Reviewer: ```suggestion
The instances of the `Path` represent the placement of the changed item within the whole entity as a prefix to the state that is either added or removed. While these paths always have the same structure, `oldPath` and `newPath` can have empty values, which represent the absence of data.
```

> When the entity is inserted in a collection, first calls to `added()` or `removed()` will have null value for the `association` argument. It represents the fact that the complete new entity is added to the collection. If the entity state is changed deeper inside its structure, the `association` is set to the association where the change occurs.

Reviewer: Maybe
```suggestion
The `association` value for `added()` and `removed()` is only provided if data is compared along associations or compositions.
```

Author: I'd rather be explicit about what null value means there.
